File: src/lira/vocabulary/role/relationship_seeder.py
"""Loads the Common Vocabulary Relationship Cache
(vocabulary/assets/common/<language>/relationships/) and seeds a
Domain's LexicalRelationship graph with it, resolving every source and
target Word against that Domain's own Dictionary -- see
vocabulary/assets/common/en/relationships/README.md for the full
seeding-order and Qualified Word resolution rationale. Relationship
assets are generated bootstrap assets; they are not the authoritative
source of lexical knowledge."""

import hashlib
import json
from pathlib import Path
from typing import List, Optional, Tuple

from lira.value_objects import Text

from ..data.dictionary import Dictionary
from ..data.lexical_relationship_store import LexicalRelationshipStore
from ..data.lexical_relationship_type import LexicalRelationshipType
from ..data.part_of_speech import PartOfSpeech
from ..data.source_reference import SourceReference
from ..data.word import Word

CATEGORY_FILES = (
    "morphological_relationships.json",
    "semantic_relationships.json",
    "orthographic_relationships.json",
)
MANIFEST_FILE = "manifest.json"

DEFAULT_ASSETS_ROOT = Path(__file__).resolve().parent.parent / "assets" / "common"

CACHE_SOURCE_REFERENCE = SourceReference(
    source_name=Text(value="LIRA English Common Vocabulary Relationship Cache v1"),
    source_version=Text(value="1.0.0"),
)

# Every seeded relationship is a curated, hand-authored linguistic fact
# ("be" -> "am" is FIRST_PERSON_FORM), not an observation or inference
# with genuine uncertainty attached -- so it gets the same
# as-good-as-certain weight the Knowledge Layer uses for directly
# authored facts elsewhere (knowledge/data/tensor_graph.py), on all
# four system properties. 0.9999 rather than a literal 1.0 by the same
# convention: certainty is never asserted as exactly 1.0.
SEEDER_DEFAULT_WEIGHT = 0.9999

# (source_lexical_form, source_part_of_speech, target_lexical_form,
# target_part_of_speech, relationship_type). The two part_of_speech
# entries are usually None -- Dictionary.lookup()'s first-seeded-wins
# default is exact and unambiguous for every closed-class/mandatory
# word, which is all a spec ever needed to disambiguate until promoted
# words started creating open-class homographs (e.g. "state" NOUN and
# VERB both promoted -- vocabulary/assets/common/en/README.md's
# Homographs table). A spec whose endpoint IS genuinely ambiguous names
# an explicit part_of_speech instead of leaving it to load order --
# see _resolve below and assets/common/en/relationships/README.md's own
# schema section.
RelationshipSpec = Tuple[str, Optional[PartOfSpeech], str, Optional[PartOfSpeech], LexicalRelationshipType]


class RelationshipSeeder:
    def __init__(self, language_code: str = "en", assets_root: Optional[Path] = None):
        self.language_code = language_code
        self.assets_dir = (assets_root or DEFAULT_ASSETS_ROOT) / language_code / "relationships"
        self._cache: Optional[List[RelationshipSpec]] = None

    def validate_assets(self) -> None:
        """Validates JSON schema, per-file and total relationship
        counts, relationship kind validity, mandatory file existence,
        manifest consistency, and the manifest checksum."""
        if not self.assets_dir.is_dir():
            raise FileNotFoundError(f"no Common Vocabulary Relationship Cache for language '{self.language_code}' at {self.assets_dir}")

        manifest_path = self.assets_dir / MANIFEST_FILE
        if not manifest_path.is_file():
            raise FileNotFoundError(f"mandatory Relationship Cache file missing: {MANIFEST_FILE}")
        manifest = json.loads(manifest_path.read_text())
        for required_key in ("schema_version", "language_code", "relationship_count", "files", "checksum"):
            if required_key not in manifest:
                raise ValueError(f"{MANIFEST_FILE}: missing required key '{required_key}'")

        computed_total = 0
        for filename in CATEGORY_FILES:
            path = self.assets_dir / filename
            if not path.is_file():
                raise FileNotFoundError(f"mandatory Relationship Cache file missing: {filename}")
            doc = json.loads(path.read_text())
            for required_key in ("schema_version", "language_code", "relationship_category", "count", "relationships"):
                if required_key not in doc:
                    raise ValueError(f"{filename}: missing required key '{required_key}'")
            if doc["count"] != len(doc["relationships"]):
                raise ValueError(f"{filename}: count {doc['count']} does not match {len(doc['relationships'])} relationship entries")
            for entry in doc["relationships"]:
                kind = entry.get("relationship_kind")
                if kind not in LexicalRelationshipType.__members__:
                    raise ValueError(f"{filename}: unknown relationship_kind '{kind}'")
                if not entry.get("source_lexical_form") or not entry.get("target_lexical_form"):
                    raise ValueError(f"{filename}: relationship entry missing source_lexical_form or target_lexical_form")
                for pos_field in ("source_part_of_speech", "target_part_of_speech"):
                    pos_value = entry.get(pos_field)
                    if pos_value is not None and pos_value not in PartOfSpeech.__members__:
                        raise ValueError(f"{filename}: unknown {pos_field} '{pos_value}'")
            computed_total += doc["count"]

        if manifest["relationship_count"] != computed_total:
            raise ValueError(
                f"manifest.json relationship_count ({manifest['relationship_count']}) "
                f"does not match the computed total ({computed_total})"
            )

        computed_checksum = self._compute_checksum()
        if manifest["checksum"] != computed_checksum:
            raise ValueError("manifest.json checksum does not match the relationship files' contents")

    def _compute_checksum(self) -> str:
        digest = hashlib.sha256()
        for filename in sorted(CATEGORY_FILES):
            digest.update((self.assets_dir / filename).read_bytes())
        return digest.hexdigest()

    def load_relationship_specs(self) -> List[RelationshipSpec]:
        """Validates the assets, then parses every category file into
        RelationshipSpec tuples. Cached after the first call."""
        if self._cache is not None:
            return list(self._cache)

        self.validate_assets()
        specs = []
        for filename in CATEGORY_FILES:
            doc = json.loads((self.assets_dir / filename).read_text())
            for entry in doc["relationships"]:
                source_pos = entry.get("source_part_of_speech")
                target_pos = entry.get("target_part_of_speech")
                specs.append((
                    entry["source_lexical_form"],
                    PartOfSpeech[source_pos] if source_pos else None,
                    entry["target_lexical_form"],
                    PartOfSpeech[target_pos] if target_pos else None,
                    LexicalRelationshipType[entry["relationship_kind"]],
                ))
        self._cache = specs
        return list(specs)

    def seed_domain(self, domain) -> int:
        """Resolves and creates every relationship in the cache against
        `domain`'s own Dictionary (Qualified Word = this Domain + a
        lexical form, never lexical form alone), skipping any that
        already exist (same source Word, kind, and target Word) and
        raising if a source or target Word cannot be resolved -- a
        cache/asset inconsistency, not something to seed around
        silently. Words must already be seeded (WordSeeder) before
        calling this. Every relationship created this way gets all four
        system properties set to SEEDER_DEFAULT_WEIGHT, not the
        LexicalRelationshipProcessor.create default of 0.0 -- a seeded
        relationship is a curated fact, not an unweighted placeholder.

        Resolution happens as a complete first pass, before any
        relationship is created: if the Nth spec in the cache can't be
        resolved, the first N-1 are never created either, rather than
        this Domain ending up with a partially-seeded relationship
        graph and no indication where it stopped."""
        dictionary: Dictionary = domain.vocabulary.dictionary
        store: LexicalRelationshipStore = domain.vocabulary.lexical_relationships
        processor = domain.vocabulary.lexical_relationship_processor

        resolved = []
        for source_form, source_pos, target_form, target_pos, relationship_type in self.load_relationship_specs():
            source_word = self._resolve(dictionary, source_form, source_pos)
            if source_word is None:
                raise ValueError(f"cannot resolve source Word '{source_form}'"
                                  f"{f' ({source_pos.name})' if source_pos else ''} in Domain '{domain.name}'")
            target_word = self._resolve(dictionary, target_form, target_pos)
            if target_word is None:
                raise ValueError(f"cannot resolve target Word '{target_form}'"
                                  f"{f' ({target_pos.name})' if target_pos else ''} in Domain '{domain.name}'")
            resolved.append((source_word, target_word, relationship_type))

        seeded = 0
        for source_word, target_word, relationship_type in resolved:
            if self._relationship_exists(store, source_word.uuid.value, target_word.uuid.value, relationship_type):
                continue

            processor.create(
                source_word_id=source_word.uuid.value,
                target_word_id=target_word.uuid.value,
                relationship_type=relationship_type,
                source_references=(CACHE_SOURCE_REFERENCE,),
                confidence=SEEDER_DEFAULT_WEIGHT,
                provenance=SEEDER_DEFAULT_WEIGHT,
                temporal=SEEDER_DEFAULT_WEIGHT,
                activation=SEEDER_DEFAULT_WEIGHT,
            )
            seeded += 1
        return seeded

    @staticmethod
    def _resolve(dictionary: Dictionary, lexical_form: str, part_of_speech: Optional[PartOfSpeech]) -> Optional[Word]:
        """Resolves one spec endpoint against `dictionary`. Without a
        part_of_speech hint, defers to Dictionary.lookup()'s own
        first-seeded-wins default, unchanged from before this method
        existed -- every spec that doesn't need disambiguating keeps
        behaving exactly as it always did. With one, resolves via
        lookup_all() and picks the matching sense, ignoring load order
        entirely -- the only way to correctly target the VERB sense of
        a lexical_form whose NOUN sense loaded first (or vice versa)."""
        if part_of_speech is None:
            return dictionary.lookup(lexical_form)
        return next((word for word in dictionary.lookup_all(lexical_form) if word.part_of_speech == part_of_speech), None)

    @staticmethod
    def _relationship_exists(store: LexicalRelationshipStore, source_word_id: str,
                              target_word_id: str, relationship_type: LexicalRelationshipType) -> bool:
        return any(
            relationship.target_word_id.value == target_word_id and relationship.relationship_type == relationship_type
            for relationship in store.outgoing(source_word_id)
        )

File: prototype/src/lira/vocabulary/role/relationship_seeder.ts
/** Loads the Common Vocabulary Relationship Cache
 * (vocabulary/assets/common/<language>/relationships/) and seeds a
 * Domain's MorphologicalPointerRelationship graph with it, resolving every source
 * and target Word against that Domain's own Dictionary. Relationship
 * assets are generated bootstrap assets; they are not the authoritative
 * source of lexical knowledge.
 *
 * Ported from vocabulary/role/relationship_seeder.py. validateAssets()
 * and everything that calls it are async here (unlike the synchronous
 * Python original) because checksum verification uses the Web Crypto
 * API's `crypto.subtle.digest`, which is promise-based. */

import type { Dictionary } from "../data/dictionary";
import { MorphologicalPointerRelationshipStore } from "../data/morphological_pointer_relationship_store";
import { LexicalRelationshipType } from "../data/enums/lexical_relationship_type";
import { PartOfSpeech } from "../data/enums/part_of_speech";
import type { Phrases } from "../data/phrases";
import type { SemanticRelationshipStore } from "../data/semantic_relationship_store";
import type { LexicalRelationshipStore } from "../data/lexical_relationship_store";
import type { SourceReference } from "../data/source_reference";
import type { Word } from "../data/entities/word";
import type { WordForms } from "../data/word_forms";
import { graphUuid } from "./word_form_processor";
import { graphUuid as wordGraphUuid } from "./word_processor";
import {
  readRelationshipFile,
  readRelationshipFileRaw,
  relationshipDirectoryExists,
  sha256Hex,
  type RelationshipFileEntry,
  type RelationshipManifestDocument,
} from "./asset_loader";
import type { MorphologicalPointerRelationshipProcessor } from "./morphological_pointer_relationship_processor";
import type { SemanticRelationshipProcessor } from "./semantic_relationship_processor";
import type { LexicalRelationshipProcessor } from "./lexical_relationship_processor";
import { LEXICAL_TO_SEMANTIC_KIND } from "./word_seeder";

export const CATEGORY_FILES = [
  "morphological_relationships.json",
  "semantic_relationships.json",
  "orthographic_relationships.json",
] as const;
const MANIFEST_FILE = "manifest.json";

const CACHE_SOURCE_REFERENCE: SourceReference = {
  sourceName: { value: "LIRA English Common Vocabulary Relationship Cache v1" },
  sourceVersion: { value: "1.0.0" },
};

// Every seeded relationship is a curated, hand-authored linguistic fact
// ("be" -> "am" is FIRST_PERSON_FORM), not an observation or inference
// with genuine uncertainty attached -- so it gets the same
// as-good-as-certain weight the Knowledge Layer uses for directly
// authored facts elsewhere. 0.9999 rather than a literal 1.0: certainty
// is never asserted as exactly 1.0.
const SEEDER_DEFAULT_WEIGHT = 0.9999;

interface RelationshipSpec {
  sourceForm: string;
  sourcePos?: PartOfSpeech;
  sourceDomainTag?: string;
  targetForm: string;
  targetPos?: PartOfSpeech;
  targetDomainTag?: string;
  relationshipType: LexicalRelationshipType;
}

export class RelationshipSeeder {
  private cache: RelationshipSpec[] | null = null;

  constructor(private readonly languageCode: string = "en") {}

  /** Validates JSON schema, per-file and total relationship counts,
   * relationship kind validity, mandatory file existence, manifest
   * consistency, and the manifest checksum. */
  async validateAssets(): Promise<void> {
    if (!relationshipDirectoryExists(this.languageCode)) {
      throw new Error(`no Common Vocabulary Relationship Cache for language '${this.languageCode}'`);
    }

    const manifest = readRelationshipFile(this.languageCode, MANIFEST_FILE) as unknown as
      | RelationshipManifestDocument
      | undefined;
    if (manifest === undefined) throw new Error(`mandatory Relationship Cache file missing: ${MANIFEST_FILE}`);

    let computedTotal = 0;
    for (const filename of CATEGORY_FILES) {
      const doc = readRelationshipFile(this.languageCode, filename);
      if (doc === undefined) throw new Error(`mandatory Relationship Cache file missing: ${filename}`);
      if (doc.count !== doc.relationships.length) {
        throw new Error(`${filename}: count ${doc.count} does not match ${doc.relationships.length} relationship entries`);
      }
      for (const entry of doc.relationships) {
        if (!(entry.relationship_kind in LexicalRelationshipType)) {
          throw new Error(`${filename}: unknown relationship_kind '${entry.relationship_kind}'`);
        }
        if (!entry.source_lexical_form || !entry.target_lexical_form) {
          throw new Error(`${filename}: relationship entry missing source_lexical_form or target_lexical_form`);
        }
        for (const posValue of [entry.source_part_of_speech, entry.target_part_of_speech]) {
          if (posValue !== undefined && posValue !== null && !(posValue in PartOfSpeech)) {
            throw new Error(`${filename}: unknown part_of_speech '${posValue}'`);
          }
        }
      }
      computedTotal += doc.count;
    }

    if (manifest.relationship_count !== computedTotal) {
      throw new Error(
        `manifest.json relationship_count (${manifest.relationship_count}) does not match the computed total (${computedTotal})`,
      );
    }

    const computedChecksum = await this.computeChecksum();
    if (manifest.checksum !== computedChecksum) {
      throw new Error("manifest.json checksum does not match the relationship files' contents");
    }
  }

  private async computeChecksum(): Promise<string> {
    const sortedFiles = [...CATEGORY_FILES].sort();
    const texts = sortedFiles.map((filename) => {
      const raw = readRelationshipFileRaw(this.languageCode, filename);
      if (raw === undefined) throw new Error(`mandatory Relationship Cache file missing: ${filename}`);
      return raw;
    });
    return sha256Hex(texts);
  }

  /** Validates the assets, then parses every category file into
   * RelationshipSpecs. Cached after the first call. */
  async loadRelationshipSpecs(): Promise<readonly RelationshipSpec[]> {
    if (this.cache !== null) return this.cache.slice();

    await this.validateAssets();
    const specs: RelationshipSpec[] = [];
    for (const filename of CATEGORY_FILES) {
      const doc = readRelationshipFile(this.languageCode, filename);
      for (const entry of doc?.relationships ?? []) specs.push(this.entryToSpec(entry));
    }
    this.cache = specs;
    return specs.slice();
  }

  private entryToSpec(entry: RelationshipFileEntry): RelationshipSpec {
    return {
      sourceForm: entry.source_lexical_form,
      sourcePos: entry.source_part_of_speech ? PartOfSpeech[entry.source_part_of_speech as keyof typeof PartOfSpeech] : undefined,
      sourceDomainTag: entry.source_domain_tag ?? undefined,
      targetForm: entry.target_lexical_form,
      targetPos: entry.target_part_of_speech ? PartOfSpeech[entry.target_part_of_speech as keyof typeof PartOfSpeech] : undefined,
      targetDomainTag: entry.target_domain_tag ?? undefined,
      relationshipType: LexicalRelationshipType[entry.relationship_kind as keyof typeof LexicalRelationshipType],
    };
  }

  /** Resolves and creates every relationship in the cache against
   * `domain`'s own Dictionary (Qualified Word = this Domain + a
   * lexical form, never lexical form alone), skipping any that already
   * exist and raising if a source or target Word cannot be resolved.
   * Words must already be seeded (WordSeeder) before calling this.
   *
   * Resolution happens as a complete first pass, before any
   * relationship is created: if the Nth spec can't be resolved, the
   * first N-1 are never created either -- unless `options.skipUnresolvable`
   * is set, in which case an unresolvable spec is simply skipped instead
   * of aborting the whole run. Default false, preserving this method's
   * original all-or-nothing behaviour (a real cache-consistency bug
   * should still fail loudly) for every existing caller. The one caller
   * that opts in is the Vocabulary view's own "Seed Vocabulary" toolbar
   * action (vocabulary_worker.ts's handleSeedCommonVocabulary), paired
   * with WordSeeder.seedDomain's own `excludeOpenClasses` -- once that
   * option leaves NOUN/VERB/ADJECTIVE/ADVERB Words unseeded by design,
   * a large share of this Common Relationship Cache's own specs (most
   * of which relate open-class words) can no longer resolve, and that's
   * expected, not a cache bug.
   *
   * A spec naming a multi-word closed-class form ("in spite of", now a
   * Phrase -- phrase.ts's own docstring) is a second, unconditional
   * exception, regardless of `skipUnresolvable`: MorphologicalPointerRelationshipStore
   * only ever connects two Word uuids, so a curated fact like "despite
   * SYNONYM in spite of" (semantic_relationships.json) has no graph
   * node to attach its Phrase-side edge to in this first pass -- silently
   * skipped, the same as any other spec this store's own shape simply
   * can't represent, not surfaced as a cache-consistency bug the way an
   * unresolvable Word genuinely would be. */
  async seedDomain(
    domain: {
      name: string;
      vocabulary: {
        dictionary: Dictionary;
        phrases: Phrases;
        morphologicalPointerRelationships: MorphologicalPointerRelationshipStore;
        morphologicalPointerRelationshipProcessor: MorphologicalPointerRelationshipProcessor;
        semanticRelationships: SemanticRelationshipStore;
        semanticRelationshipProcessor: SemanticRelationshipProcessor;
        wordForms?: WordForms;
        lexicalRelationships?: LexicalRelationshipStore;
        lexicalRelationshipProcessor?: LexicalRelationshipProcessor;
      };
    },
    options?: { skipUnresolvable?: boolean },
  ): Promise<number> {
    const skipUnresolvable = options?.skipUnresolvable ?? false;
    const dictionary = domain.vocabulary.dictionary;
    const phraseBook = domain.vocabulary.phrases;
    const store = domain.vocabulary.morphologicalPointerRelationships;
    const processor = domain.vocabulary.morphologicalPointerRelationshipProcessor;
    const semanticProcessor = domain.vocabulary.semanticRelationshipProcessor;
    const wordForms = domain.vocabulary.wordForms;
    const lexicalProcessor = domain.vocabulary.lexicalRelationshipProcessor;
    const lexicalExistingEdges = new Set<string>();
    if (domain.vocabulary.lexicalRelationships !== undefined) {
      for (const relationship of domain.vocabulary.lexicalRelationships.all()) {
        lexicalExistingEdges.add(
          `${relationship.sourceWordFormId.value}|${relationship.sourceSenseId.value}|${relationship.targetWordFormId.value}|${relationship.targetSenseId.value}|${relationship.relationshipType}`,
        );
      }
    }

    const resolved: Array<[Word, Word, LexicalRelationshipType]> = [];
    for (const spec of await this.loadRelationshipSpecs()) {
      const sourceWord = this.resolve(dictionary, spec.sourceForm, spec.sourcePos, spec.sourceDomainTag);
      if (sourceWord === undefined) {
        if (skipUnresolvable || this.isPhraseOnly(phraseBook, spec.sourceForm, spec.sourcePos)) continue;
        throw new Error(
          `cannot resolve source Word '${spec.sourceForm}'` +
            (spec.sourcePos !== undefined ? ` (${PartOfSpeech[spec.sourcePos]})` : "") +
            (spec.sourceDomainTag ? ` [${spec.sourceDomainTag}]` : "") +
            ` in Domain '${domain.name}'`,
        );
      }
      const targetWord = this.resolve(dictionary, spec.targetForm, spec.targetPos, spec.targetDomainTag);
      if (targetWord === undefined) {
        if (skipUnresolvable || this.isPhraseOnly(phraseBook, spec.targetForm, spec.targetPos)) continue;
        throw new Error(
          `cannot resolve target Word '${spec.targetForm}'` +
            (spec.targetPos !== undefined ? ` (${PartOfSpeech[spec.targetPos]})` : "") +
            (spec.targetDomainTag ? ` [${spec.targetDomainTag}]` : "") +
            ` in Domain '${domain.name}'`,
        );
      }
      resolved.push([sourceWord, targetWord, spec.relationshipType]);
    }

    let seeded = 0;
    const semanticExistingEdges = new Set<string>();
    for (const [sourceWord, targetWord, relationshipType] of resolved) {
      if (!this.relationshipExists(store, wordGraphUuid(sourceWord), wordGraphUuid(targetWord), relationshipType)) {
        processor.create({
          sourceWordId: wordGraphUuid(sourceWord),
          targetWordId: wordGraphUuid(targetWord),
          relationshipType,
          sourceReferences: [CACHE_SOURCE_REFERENCE],
          confidence: SEEDER_DEFAULT_WEIGHT,
          provenance: SEEDER_DEFAULT_WEIGHT,
          temporal: SEEDER_DEFAULT_WEIGHT,
          activation: SEEDER_DEFAULT_WEIGHT,
        });
        seeded += 1;
      }

      this.copyOntoPermanentModel(
        semanticProcessor,
        semanticExistingEdges,
        lexicalProcessor,
        lexicalExistingEdges,
        wordForms,
        sourceWord,
        targetWord,
        relationshipType,
      );
    }
    return seeded;
  }

  /** Every curated fact this cache seeds also lands on its own permanent
   * home -- the target's own base-lemma WordForm's own `contractionOf`
   * for CONTRACTION (WordForm's own docstring, data/entities/word_form.ts, on why
   * it's a WordForm-level fact and many-to-many)
   * or a genuine SemanticRelationship for a true sense-to-sense semantic
   * kind (LEXICAL_TO_SEMANTIC_KIND's own docstring, word_seeder.ts, on
   * exactly which those are and why this cache uses four kinds --
   * SYNONYM/HYPONYM/HOLONYM -- WordNet-seeded data never actually
   * produces) -- the same split WordSeeder.seedWordNet's own
   * copySemanticRelationship applies to WordNet-sourced facts, applied
   * here to hand-curated ones instead. Each curated Word gets exactly
   * one private Sense of its own (registerUniqueSense, word_seeder.ts),
   * so `senseIds[0]` is never the "arbitrary primary sense of several"
   * simplification it would be for a genuinely polysemous WordNet Word --
   * there's only ever the one. A Word with no Sense at all yet (this
   * cache runs after WordSeeder.seedDomain's own Word seeding, but a
   * caller that skips Sense registration entirely is defensively
   * possible) silently contributes nothing rather than throwing --
   * missing Sense data is a Word-seeding gap, not something this cache
   * seeding pass should surface as its own error. */
  private copyOntoPermanentModel(
    semanticProcessor: SemanticRelationshipProcessor,
    semanticExistingEdges: Set<string>,
    lexicalProcessor: LexicalRelationshipProcessor | undefined,
    lexicalExistingEdges: Set<string>,
    wordForms: WordForms | undefined,
    sourceWord: Word,
    targetWord: Word,
    relationshipType: LexicalRelationshipType,
  ): void {
    if (relationshipType === LexicalRelationshipType.CONTRACTION) {
      // contractionOf lives on the target's own base-lemma WordForm now
      // (WordForm's own docstring on why), not on Word -- `wordForms`
      // undefined (defensively possible, this method's own docstring)
      // means there's nowhere left to record this fact at all, the same
      // "silently contributes nothing" outcome a missing Sense already
      // gets just below.
      const targetForm = wordForms?.registerBaseLemmaForm(targetWord);
      const sourceWordUuid = wordGraphUuid(sourceWord);
      if (targetForm !== undefined && !targetForm.contractionOf.some((id) => id.value === sourceWordUuid)) {
        targetForm.contractionOf = [...targetForm.contractionOf, { value: sourceWordUuid }];
      }
    }

    const semanticKind = LEXICAL_TO_SEMANTIC_KIND[relationshipType];
    if (semanticKind !== undefined) {
      const sourceSenseId = wordForms?.senseIdsOf(sourceWord)[0];
      const targetSenseId = wordForms?.senseIdsOf(targetWord)[0];
      if (sourceSenseId !== undefined && targetSenseId !== undefined && sourceSenseId.value !== targetSenseId.value) {
        const key = `${sourceSenseId.value}|${targetSenseId.value}|${semanticKind}`;
        if (!semanticExistingEdges.has(key)) {
          semanticExistingEdges.add(key);
          semanticProcessor.create({
            sourceSenseId: sourceSenseId.value,
            targetSenseId: targetSenseId.value,
            relationshipType: semanticKind,
            sourceReferences: [CACHE_SOURCE_REFERENCE],
            confidence: SEEDER_DEFAULT_WEIGHT,
            provenance: SEEDER_DEFAULT_WEIGHT,
            temporal: SEEDER_DEFAULT_WEIGHT,
            activation: SEEDER_DEFAULT_WEIGHT,
          });
        }
      }
      return;
    }

    // Every Morphological/Orthographic-group kind (including
    // CONTRACTION -- falls through from the branch above rather than
    // returning early there, since a curated contraction is *both* a
    // Word.contractionOf fact and its own permanent LexicalRelationship
    // now) becomes a real, permanent LexicalRelationship --
    // word_seeder.ts's own copyLexicalRelationship() applied here to
    // hand-curated facts instead of WordNet-sourced ones. Each curated
    // Word has exactly one private Sense (registerUniqueSense's own
    // "one per entry" contract), so senseIds[0] is unambiguous, not an
    // arbitrary-primary-of-several simplification.
    if (lexicalProcessor === undefined || wordForms === undefined) return;
    const sourceSenseId = wordForms.senseIdsOf(sourceWord)[0];
    const targetSenseId = wordForms.senseIdsOf(targetWord)[0];
    if (sourceSenseId === undefined || targetSenseId === undefined || sourceSenseId.value === targetSenseId.value) return;
    const sourceForm = wordForms.registerBaseLemmaForm(sourceWord);
    const targetForm = wordForms.registerBaseLemmaForm(targetWord);
    const sourceFormUuid = graphUuid(sourceForm);
    const targetFormUuid = graphUuid(targetForm);
    const key = `${sourceFormUuid}|${sourceSenseId.value}|${targetFormUuid}|${targetSenseId.value}|${relationshipType}`;
    if (lexicalExistingEdges.has(key)) return;
    lexicalExistingEdges.add(key);
    lexicalProcessor.create({
      sourceWordFormId: sourceFormUuid,
      sourceSenseId: sourceSenseId.value,
      targetWordFormId: targetFormUuid,
      targetSenseId: targetSenseId.value,
      relationshipType,
      sourceReferences: [CACHE_SOURCE_REFERENCE],
      confidence: SEEDER_DEFAULT_WEIGHT,
      provenance: SEEDER_DEFAULT_WEIGHT,
      temporal: SEEDER_DEFAULT_WEIGHT,
      activation: SEEDER_DEFAULT_WEIGHT,
    });
  }

  /** Resolves one spec endpoint against `dictionary`. Without a
   * partOfSpeech hint, defers to Dictionary.lookup()'s own
   * first-seeded-wins default. With one, resolves via lookupAll() and
   * picks the matching sense, ignoring load order entirely. With a
   * domainTag too, also requires the candidate's own domainTag to
   * match. */
  private resolve(dictionary: Dictionary, lexicalForm: string, partOfSpeech?: PartOfSpeech, domainTag?: string): Word | undefined {
    if (partOfSpeech === undefined) return dictionary.lookup(lexicalForm);
    const candidates = dictionary.lookupAll(lexicalForm).filter((word) => word.partOfSpeech === partOfSpeech);
    return candidates.find((word) => (word.domainTag?.value ?? undefined) === domainTag);
  }

  /** Whether `lexicalForm` (optionally narrowed by `partOfSpeech`)
   * names a real, seeded Phrase -- seedDomain()'s own signal to skip a
   * spec silently rather than treat a Dictionary miss as a cache bug.
   * A Phrase endpoint genuinely can carry a MorphologicalPointerRelationship now
   * (word_seeder.ts's own seedWordNet does exactly that), but this
   * class's own `resolve()` only ever looks a spec's source/target Word
   * up in `dictionary` -- the bundled Common Relationship Cache this
   * class seeds from is hand-curated Word-to-Word data, a different
   * source from WordNet, with no spec of its own naming a Phrase
   * endpoint on purpose. */
  private isPhraseOnly(phraseBook: Phrases, lexicalForm: string, partOfSpeech?: PartOfSpeech): boolean {
    const candidates = phraseBook.lookupAll(lexicalForm);
    return partOfSpeech === undefined ? candidates.length > 0 : candidates.some((phrase) => phrase.partOfSpeech === partOfSpeech);
  }

  private relationshipExists(
    store: MorphologicalPointerRelationshipStore,
    sourceWordId: string,
    targetWordId: string,
    relationshipType: LexicalRelationshipType,
  ): boolean {
    return store
      .outgoing(sourceWordId)
      .some((relationship) => relationship.targetWordId.value === targetWordId && relationship.relationshipType === relationshipType);
  }
}
